### PR TITLE
Ignore subdomain for soa queries

### PR DESCRIPTION
**Pull request by kesselb**

### CODE REVIEW

File: lib/private/Http/Client/DnsPinMiddleware.php [hunk 1]
```diff
@@ -49,24 +65,19 @@ private function dnsResolve(string $target, int $recursionCount) : array {
...
+		foreach ($dnsTypes as $dnsType) {
 			if ($this->negativeDnsCache->isNegativeCached($target, $dnsType)) {
-				unset($dnsTypes[$key]);
 				continue;
 			}
 
 			$dnsResponses = dns_get_record($target, $dnsType);
 			$canHaveCnameRecord = true;
 			if (count($dnsResponses) > 0) {
-				foreach ($dnsResponses as $key => $dnsResponse) {
```
Comment (kesselb) [inline lib/private/Http/Client/DnsPinMiddleware.php:69]: Code cleanup: As $key is unused and may overwrite $key from the outer loop.

File: lib/private/Http/Client/DnsPinMiddleware.php [hunk 2]
```diff
@@ -49,24 +65,19 @@ private function dnsResolve(string $target, int $recursionCount) : array {
...
-			$dnsNegativeTtl = null;
-		}
+		$soaDnsEntry = $this->soaRecord($target);
+		$dnsNegativeTtl = $soaDnsEntry[0]['minimum-ttl'] ?? null;
 
 		$dnsTypes = [DNS_A, DNS_AAAA, DNS_CNAME];
-		foreach ($dnsTypes as $key => $dnsType) {
+		foreach ($dnsTypes as $dnsType) {
 			if ($this->negativeDnsCache->isNegativeCached($target, $dnsType)) {
-				unset($dnsTypes[$key]);
```
Comment (kesselb) [inline lib/private/Http/Client/DnsPinMiddleware.php:62]: As $dnsTypes is not used it's fine to not unset the value and just continue.

File: lib/private/Http/Client/DnsPinMiddleware.php [hunk 3]
```diff
@@ -41,6 +41,22 @@ public function __construct(
...
+	 * @return array|false
+	 */
+	private function soaRecord(string $target) {
+		$labels = explode('.', $target);
+
+		$top = count($labels) >= 2 ? array_pop($labels) : '';
+		$second = array_pop($labels);
+
+		$hostname = $second . '.' . $top;
+		return dns_get_record($hostname, DNS_SOA);
```
Comment (kesselb) [inline lib/private/Http/Client/DnsPinMiddleware.php:57]:
```
$ host -t SOA www.nextcloud.com
www.nextcloud.com has no SOA record
```
```
$ host -t SOA nextcloud.com
nextcloud.com has SOA record ns.inwx.de. hostmaster.inwx.de. 2021033001 10800 3600 604800 3600
```

**Review by LukasReschke — APPROVED** [review]

**Comment by LukasReschke** [comment]
phpcs is complaining

File: lib/private/Http/Client/DnsPinMiddleware.php [hunk 4]
```diff
@@ -49,24 +65,19 @@ private function dnsResolve(string $target, int $recursionCount) : array {
...
 		$targetIps = [];
 
-		$soaDnsEntry = dns_get_record($target, DNS_SOA);
-		if (isset($soaDnsEntry[0]) && isset($soaDnsEntry[0]['minimum-ttl'])) {
-			$dnsNegativeTtl = $soaDnsEntry[0]['minimum-ttl'];
-		} else {
-			$dnsNegativeTtl = null;
-		}
+		$soaDnsEntry = $this->soaRecord($target);
+		$dnsNegativeTtl = $soaDnsEntry[0]['minimum-ttl'] ?? null;
```
Comment (blizzz) [inline lib/private/Http/Client/DnsPinMiddleware.php:69]: according to the php doc `$soaDnsEntry` can be false. Perhaps you can turn false to null and add it as return type hint.

Comment (kesselb) [reply lib/private/Http/Client/DnsPinMiddleware.php]: done

**Review by Pytal — APPROVED** [review]

**Pull request merged**